### DISCUSSION TITLE
How refs work Popover abstracting example

### DISCUSSION THREAD

**Discussion by jonrimmer** [ariakit/ariakit#896] [Questions] [answered]
I'm confused about the approach used for refs in the Popover abstracting example in the docs: https://reakit.io/docs/popover/#abstracting

The `Popover` component has a `disclosure` render prop, but you then access `disclosure.ref`, even though the element passed into the example has no `ref` assigned to it. I've never seen this pattern before, and it doesn't seem to match any in the React docs. What exactly is happening? How can you safely access `ref` if there's no `useRef()` or anything assigned to it? And if I'm implementing this in TypeScript, how should I type it? Usually I type render props using `ReactNode`, or perhaps `ReactElement`, but if I do so it complains that these have no accessible `ref` property.

**Comment by diegohaz** [comment, marked as answer]
Just to make the distinction clear, that's not a render prop (that is, it's not a function). The `disclosure.ref` is there so you can assign a ref to the element passed to the `disclosure` prop, for example:
```jsx
function Example() {
  const ref = React.useRef();
  React.useEffect(() => console.log(ref), []);
  return (
    <Popover
      aria-label="Custom popover"
      disclosure={<button ref={ref}>Open custom popover</button>}
    >
      Custom popover
    </Popover>
  );
}
```

If you don't pass `disclosure.ref` to `PopoverDisclosure`, this ref will be ignored.

You can type it as `ReactElement & RefAttributes<HTMLElement>`.
